File: packages/dart/lib/src/catalogs/banxico/cetes_28.dart
/// CETES 28 días Catalog
///
/// Catalog of CETES 28-day values from Banco de México
library;

/// CETES 28-day Catalog
class Cetes28Catalog {
  static List<Map<String, dynamic>>? _data;

  static void _loadData() {
    if (_data != null) return;
    // TODO: Load from packages/shared-data/banxico/cetes_28.json
    _data = [];
  }

  static List<Map<String, dynamic>> getData() {
    _loadData();
    return List.from(_data!);
  }

  static Map<String, dynamic>? getActual() {
    _loadData();
    if (_data!.isEmpty) return null;
    final sorted = List.from(_data!)
      ..sort((a, b) => (b['fecha'] as String).compareTo(a['fecha'] as String));
    return sorted.first;
  }

  static double? getTasaActual() {
    final record = getActual();
    return record?['tasa'] as double?;
  }
}

File: packages/typescript/src/catalogs/banxico/cetes-28.ts
/**
 * CETES 28 días Catalog
 *
 * Provides access to CETES 28-day values from Banco de México.
 */

import cetesData from '../../../../shared-data/banxico/cetes_28.json';

export interface CETES28Record {
  fecha: string;
  tasa: number;
  plazo_dias: number;
  instrumento: string;
  tipo: string;
  año: number;
  mes: number;
  fuente: string;
}

export class CETES28Catalog {
  private static data: CETES28Record[] = cetesData as CETES28Record[];
  private static byFecha: Map<string, CETES28Record> | null = null;

  private static _loadIndexes(): void {
    if (this.byFecha !== null) return;
    this.byFecha = new Map();
    for (const record of this.data) {
      this.byFecha.set(record.fecha, record);
    }
  }

  static getData(): CETES28Record[] {
    return [...this.data];
  }

  static getPorFecha(fecha: string): CETES28Record | null {
    this._loadIndexes();
    return this.byFecha!.get(fecha) ?? null;
  }

  static getActual(): CETES28Record | null {
    if (this.data.length === 0) return null;
    return this.data.reduce((latest, record) => (record.fecha > latest.fecha ? record : latest));
  }

  static getTasaActual(): number | null {
    return this.getActual()?.tasa ?? null;
  }
}

export function getCETESActual(): CETES28Record | null {
  return CETES28Catalog.getActual();
}
